lib/journaled/abstract_adapter_ext.rb

```diff
@@ -0,0 +1,68 @@
+module Journaled
+  module AbstractAdapterExt
+    def self.included(klass)
+      klass.set_callback(:checkout, :before) do
+        @_journaled_transaction_handler = nil
+      end
+    end
+
+    def _journaled_transaction_handler
+      if @_journaled_transaction_handler&.active?
+        @_journaled_transaction_handler
+      else
+        @_journaled_transaction_handler = TransactionHandler.new(connection: self)
+      end
+    end
+
+    class TransactionHandler
```

Comment:
This is similar to what the [`after_commit_everywhere` gem](https://github.com/Envek/after_commit_everywhere) does, but in our case we want it to be _before_ the commit, and we also want to be able to build up a list of staged events over the course of the transaction, so we couldn't actually drop that gem in as a dependency.